nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_prototype.rs

```diff
@@ -3247,13 +3247,80 @@ impl ArrayPrototype {
         Ok(ArrayIterator::from_object(agent, o, CollectionIteratorKind::Value).into_value())
     }
 
+    /// ### [23.1.3.39 Array.prototype.with ( index, value )](https://tc39.es/ecma262/#sec-array.prototype.with)
     fn with(
-        _agent: &mut Agent,
-        _gc: GcScope<'_, '_>,
-        _this_value: Value,
-        _: ArgumentsList,
+        agent: &mut Agent,
+        mut gc: GcScope<'_, '_>,
+        this_value: Value,
+        arguments: ArgumentsList,
     ) -> JsResult<Value> {
-        todo!();
+        // Fast path: Array is dense and contains no descriptors. No JS
+        // functions can thus be called by with.
+        if let Value::Array(array) = this_value {
+            let len = array.len(agent) as i64;
+            let index = arguments.get(0);
+            let relative_index =
```

Comment:
**suggestion:** Let's conditionalise the fast path with the `index` parameter being an Integer as well so we can avoid the `to_integer_or_infinity` call and the possibility of GC it triggers:

```rs
let index = arguments.get(0);
let value = arguments.get(1);
if let (Value::Array(array), Value::Integer(index)) = (this_value, index) {
  // ...
}
```